### PR TITLE
TypeScript Refactor

### PR DESCRIPTION
**Pull request by NicholasLYang**
Okay, so with JSS v10, typings are significantly improved. This isn't complete yet, but I basically turned on errors with TS and started fixing the crapload of errors. If you want to give yourself a nice scare, run this and count the errors.

### CODE REVIEW

File: src/js/modules/core/components/APIList.tsx [hunk 1]
```diff
@@ -1,49 +1,40 @@
...
     maxWidth: "500px"
   }
 };
-  
-  class APIList extends React.Component<JssRules> {
-    render (){
-      return(
+interface Props extends WithStyles<typeof styles> {}
+
+const APIList: React.SFC<Props> = () => {
```
Comment (themichaelyang) [inline src/js/modules/core/components/APIList.tsx]: Looks like the API List snuck its way back in again

Comment (NicholasLYang) [reply src/js/modules/core/components/APIList.tsx]: Deleted now!

**Comment by NicholasLYang** [comment]
Can we merge? I'd like to get this in ASAP

File: src/js/modules/core/components/SchoolInput.tsx [hunk 2]
```diff
@@ -80,30 +72,30 @@ const menuStyle = {
   fontSize: "90%",
   top: "50px", // height of your input
   left: 0,
-  position: "absolute",
+  position: "absolute" as "absolute",
```
Comment (themichaelyang) [inline src/js/modules/core/components/SchoolInput.tsx]: what the `as` do? (also line 83)

File: src/js/modules/core/components/TrackInfo.tsx [hunk 3]
```diff
@@ -107,7 +94,7 @@ const TrackInfo: React.SFC<Props> = ({ classes }) => {
         <Track
           id={0}
           key={0}
-          name="Heath & Well-Being"
+          name="Health & Well-Being"
```
Comment (themichaelyang) [inline src/js/modules/core/components/TrackInfo.tsx]: lol

File: src/js/modules/core/components/UpdatePasswordForm.tsx [hunk 4]
```diff
@@ -29,36 +26,41 @@ const UpdatePasswordForm: React.SFC<Props> = ({
...
+        if (!password) {
           errors.password = "Password is required";
-        }
-        else if (password.length < 8) {
+        } else if (password.length < 8) {
           errors.password = "Password must be at least 8 characters";
         }
 
-        if(!passwordConfirmation){
+        if (!passwordConfirmation) {
```
Comment (themichaelyang) [inline src/js/modules/core/components/UpdatePasswordForm.tsx]: this is an eagle eye catch! or did you prettify?

File: src/js/modules/core/components/UploadResumeButton.tsx [hunk 5]
```diff
@@ -46,22 +40,23 @@ const styles = (theme: Theme): UploadButtonStyles<JssRules> => ({
...
+    if (this.fileUploader) {
+      this.fileUploader.current.click();
+    }
   };
 
   handleUpload = () => {
     const { uid, uploadResume } = this.props;
     const file = this.fileUploader.current.files[0];
-    uploadResume(uid, file).then((timestamp) => console.log(timestamp));
+    uploadResume(uid, file);
```
Comment (themichaelyang) [inline src/js/modules/core/components/UploadResumeButton.tsx]: this reminds me. I think the upload resume button doesn't currently update its timestamp after an upload until a page refresh (let me confirm this)

File: tsconfig.json [hunk 6]
```diff
@@ -1,7 +1,9 @@
 {
   "compilerOptions": {
     "noImplicitAny": true,
-    "noEmit": true,
+    "sourceMap": true,
+//    "noEmit": true,
```
Comment (themichaelyang) [inline tsconfig.json]: remove comment (is this why TS wasn't shouting at us?)

File: webpack.common.js [hunk 7]
```diff
@@ -8,8 +8,9 @@ module.exports = {
       {
         test: /\.(jsx?$)|(tsx?$)/,
         exclude: /(node_modules|bower_components)/,
-        use: {
-          loader: "babel-loader"
+        loader: "ts-loader",
+        options: {
+          ignoreDiagnostics: [ 2322]
```
Comment (themichaelyang) [inline webpack.common.js]: what's this?

**Comment by stefancherubin** [comment]
Everything here looks good (sorry, I've been pretty busy today with a phone interview)

File: tsconfig.json (hunk 6 above)
Comment (NicholasLYang) [reply tsconfig.json]: Nah, that was cause we were using babel's TS plugin and not the compiler directly

File: webpack.common.js (hunk 7 above)
Comment (NicholasLYang) [reply webpack.common.js]: There's a persistent error that I couldn't fix that I'll get around to. Mostly to stop the errors from filling up the screen.

File: src/js/modules/core/components/SchoolInput.tsx (hunk 2 above)
Comment (NicholasLYang) [reply src/js/modules/core/components/SchoolInput.tsx]: This is a stupid little bug. In TS, you can type things as literal values. This means that the type of a variable can be `true`.

Basically the styles typings are done with these literal types, which means that the only typesafe options for position are the strings "absolute","relative", "fixed", etc. For some reason, this string is typed as a string and not as a string literal, so we need to cast it to the string "absolute"

File: src/js/modules/core/components/UpdatePasswordForm.tsx (hunk 4 above)
Comment (NicholasLYang) [reply src/js/modules/core/components/UpdatePasswordForm.tsx]: Probably prettier

**Pull request merged**